src/charm.py

```diff
@@ -70,27 +75,32 @@ def apply_crds(self):
         self.kubectl("apply", "-f", "templates/crd.yaml")
 
     def apply_grafana_agent(self, remote_endpoints):
-        patch_res = [
-            {"kind": "deployment", "name": "kube-ovn-monitor", "port": 10661},
-            {"kind": "daemonset", "name": "kube-ovn-pinger", "port": 8080},
-            {"kind": "daemonset", "name": "kube-ovn-cni", "port": 10665},
-        ]
         if not self.stored.prometheus_patched:
-            self.patch_prometheus_resources(patch_res)
+            patch_res = [
+                {"kind": "deployment", "name": "kube-ovn-monitor", "port": 10661},
+                {"kind": "daemonset", "name": "kube-ovn-pinger", "port": 8080},
+                {"kind": "daemonset", "name": "kube-ovn-cni", "port": 10665},
+            ]
```

Comment:
Maybe make this a constant `GRAFANA_RESOURCES` 

the constant can be the same `List[Dict]` set.  Jinja2 doesn't care if you give extra arguments in the context during the rendering. so you might be able to

If it's a constant you wouldn't have to define it in `apply_grafana_agent` and `remove_grafana_agent`

Reply:
If it was a constant, you could even push it into `patch_prometheus_resources`